tests/custom_lookups/tests.py

```diff
@@ -647,3 +671,84 @@ def test_subquery_usage(self):
+        with register_lookup(Company.place, CustomStartsWith):
+            self.assertEqual(Company.name.field.get_lookup("sw"), None)
+            self.assertEqual(Company.place.field.get_lookup("sw"), CustomStartsWith)
+
+    def test_get_lookup_from_proxy_method(self):
+        with register_lookup(models.CharField, CustomStartsWith):
+            self.assertEqual(Company.name.get_lookup("sw"), CustomStartsWith)
+            self.assertEqual(Company.place.get_lookup("sw"), CustomStartsWith)
+        with register_lookup(Company.place, CustomStartsWith):
+            self.assertEqual(Company.name.get_lookup("sw"), None)
+            self.assertEqual(Company.place.get_lookup("sw"), CustomStartsWith)
+
+    def test_get_transform_from_proxy_method(self):
+        with register_lookup(models.CharField, Div3Transform):
+            self.assertEqual(Company.name.get_transform("div3"), Div3Transform)
+            self.assertEqual(Company.place.get_transform("div3"), Div3Transform)
+        with register_lookup(Company.place, Div3Transform):
+            self.assertEqual(Company.name.get_transform("div3"), None)
+            self.assertEqual(Company.place.get_transform("div3"), Div3Transform)
+
+    def test_class_lookup_with_filter(self):
+        with register_lookup(models.CharField, CustomStartsWith):
+            self.assertCountEqual(
+                Company.objects.filter(name__sw="a"), [self.obj1, self.obj2, self.obj4]
+            )
+            self.assertCountEqual(Company.objects.filter(name__sw="c"), [self.obj3])
+
+    def test_instance_lookup_with_filter(self):
+        with register_lookup(Company.name, StartsWith4Char):
+            self.assertCountEqual(
+                Company.objects.filter(name__swf="a"), [self.obj1, self.obj2]
+            )
+            self.assertCountEqual(Company.objects.filter(name__swf="c"), [self.obj3])
+            with self.assertRaises(FieldError):
+                Company.objects.filter(place__swf="c")
+
+    def test_instance_overide_class_lookups_with_filter(self):
+        with register_lookup(models.CharField, CustomStartsWith, lookup_name="start"):
+            with register_lookup(Company.place, StartsWith4Char, lookup_name="start"):
+                self.assertCountEqual(
+                    Company.objects.filter(place__start="f"),
+                    [self.obj1, self.obj3],
+                )
+                self.assertCountEqual(
+                    Company.objects.filter(name__start="a"),
+                    [self.obj1, self.obj2, self.obj4],
+                )
+            self.assertCountEqual(
+                Company.objects.filter(place__start="f"),
+                [self.obj1, self.obj3, self.obj4],
+            )
+
+    def test_instance_overide_instance_lookup(self):
+        with register_lookup(Company.name, CustomStartsWith, lookup_name="start"):
+            Company.name.register_lookup(StartsWith4Char, lookup_name="start")
```

Comment:
Should this override an instance lookup? As far as I'm aware "instance" lookups should always have a precedence as lookups with bigger granularity :thinking: (\cc @charettes).

Reply:
Not sure I understand, aren't these both instance lookups. Isn't that equivalent of doing

```python
Company.name.register_lookup(CustomStartsWith, lookup_name="start")
Company.name.register_lookup(StartsWith4Char, lookup_name="start")
```

And thus overrides should be working?

Reply:
Ahh, yes. I missed that both are instance lookups :facepalm:

Reply:
> Not sure I understand, aren't these both instance lookups. Isn't that equivalent of doing
> 
> ```python
> Company.name.register_lookup(CustomStartsWith, lookup_name="start")
> Company.name.register_lookup(StartsWith4Char, lookup_name="start")
> ```
> 
> And thus overrides should be working?

Yeah. But wouldn't this make the test "non-isolated" if an error arises? The register_lookup context processor is used here to make sure the registered lookups are unregistered even though the test fails in between.